examples/imgui_impl_android.cpp

```diff
@@ -0,0 +1,166 @@
+//  2020-09-13: Support for Unicode characters
+//  2020-08-31: On-screen and physical keyboard input (ASCII characters only)
+//  2020-03-02: basic draft, touch input
+
+#include "imgui.h"
+#include "imgui_impl_android.h"
+#include <ctime>
+#include <map>
+#include <queue>
+
+// Android
+#include <android/native_window.h>
+#include <android/input.h>
+#include <android/keycodes.h>
+#include <android/log.h>
+
+static double                                   g_Time = 0.0;
+static ANativeWindow*                           g_Window;
+static char                                     g_logTag[] = "ImguiExample";
+static std::map<int32_t, std::queue<int32_t>>   g_keyEventQueues;
+
+int32_t ImGui_ImplAndroid_handleInputEvent(AInputEvent *inputEvent)
+{
+    ImGuiIO &io = ImGui::GetIO();
+    int32_t evType = AInputEvent_getType(inputEvent);
+    switch (evType)
+    {
+    case AINPUT_EVENT_TYPE_KEY:
+    {
+        int32_t evKeyCode = AKeyEvent_getKeyCode(inputEvent);
+        int32_t evAction = AKeyEvent_getAction(inputEvent);
+        int32_t evMetaState = AKeyEvent_getMetaState(inputEvent);
+
+        io.KeyCtrl = ((evMetaState & AMETA_CTRL_ON) != 0);
+        io.KeyShift = ((evMetaState & AMETA_SHIFT_ON) != 0);
+        io.KeyAlt = ((evMetaState & AMETA_ALT_ON) != 0);
+
+        switch (evAction)
+        {
+        // todo: AKEY_EVENT_ACTION_DOWN and AKEY_EVENT_ACTION_UP occur at once
+        // as soon as a touch pointer goes up from a key. We use a simple key event queue
+        // and process one event per key per ImGui frame in ImGui_ImplAndroid_NewFrame().
+        // ...or consider ImGui IO queue, if suitable: https://github.com/ocornut/imgui/issues/2787
+        case AKEY_EVENT_ACTION_DOWN:
+        case AKEY_EVENT_ACTION_UP:
+            g_keyEventQueues[evKeyCode].push(evAction);
+            break;
+        default:
+            break;
+        }
+        break;
+    }
+    case AINPUT_EVENT_TYPE_MOTION:
+    {
+        int32_t evAction = AMotionEvent_getAction(inputEvent);
+        int32_t evPointerIndex = (evAction & AMOTION_EVENT_ACTION_POINTER_INDEX_MASK) >> AMOTION_EVENT_ACTION_POINTER_INDEX_SHIFT;
+        int32_t evPointerId = AMotionEvent_getPointerId(inputEvent, evPointerIndex);
+        evAction &= AMOTION_EVENT_ACTION_MASK;
+        switch (evAction)
+        {
+        case AMOTION_EVENT_ACTION_DOWN:
+        case AMOTION_EVENT_ACTION_UP:
+            io.MouseDown[0] = (evAction == AMOTION_EVENT_ACTION_DOWN) ? true : false;
+            // intended fallthrough...
+        case AMOTION_EVENT_ACTION_MOVE:
+            io.MousePos = ImVec2(
+                AMotionEvent_getX(inputEvent, evPointerIndex),
+                AMotionEvent_getY(inputEvent, evPointerIndex));
+            break;
+        default:
+            break;
+        }
+    }
+        return 1;
+    default:
+        break;
+    }
+
+    return 0;
+}
+
+bool ImGui_ImplAndroid_Init(ANativeWindow *window)
+{
+    g_Window = window;
```

Comment:
Calling ANativeWindow_acquire() (and ANativeWindow_release in ImGui_ImplAndroid_Shutdown) may be a good idea. Otherwise window handle may be destroyed which may lead to crash as it tend to be destroyed when user minimize app.

What is your experience? Does app survive switching context without trouble?

Reply:
I haven't found a way of not having the window destroyed, when the app goes out of scope. The app does survive though, because we shutdown and (re)init every time the app goes out of scope, and returns.

I think one should rely on saving and loading the app state in this case. But I will check if `ANativeWindow_acquire()` and `ANativeWindow_release()` can help to keep the context alive, in cases where the system does not kill the app in the background, anyway.

Reply:
I had used them to make sure handle is valid until I have a change to handle proper message.
I didn't though of holding lock whole time. Actually I'm interested how that will go.

Reply:
Discussed again, below.